In [3]:
import os
import rasterio

In [4]:
def my_favorite_rasters():
    
    my_inputs={}
    # hardcoding some paths to files
    root = '/wsefs/pipeline/data_test_normToGrid'
    lst_raster = os.path.join(root, 'LS_eVSE_TEMP.2022.284-293.1KM.LST_TEMP.006.2022294042535.tif')
    ndvi_raster = os.path.join(root, 'GN_eVSH_NDVI.2022.284-293.1KM.VI_NDVI.001.2022310010552.tif')
    ndwi_raster = os.path.join(root, 'GN_eVSH_NDWI.2022.284-293.1KM.WI_NDWI.006.2022310173428.tif')
    ndvi_qa = os.path.join(root, 'GN_eVSH_NDVI.2022.284-293.1KM.VI_QUAL.001.2022310010552.tif')
    ndwi_qa = os.path.join(root, 'GN_eVSH_NDWI.2022.284-293.1KM.WI_QUAL.006.2022310173428.tif')


    s3_img_root = 's3://ws-in/ssebop/'
    dt_raster = os.path.join(s3_img_root, 'dT/dT_102.tif')
    tmax_raster = os.path.join(s3_img_root, 'air_temp/tmax_102.tif')
    etr_raster = os.path.join(s3_img_root, 'etr/Hobbins/etr_102.tif')
    global_water_raster = os.path.join(s3_img_root, 'masks/global_water_mask_inland.tif')
    
    # print('WHAT about ndwi_qa do we need that?')
    
    
    my_inputs['gold_standard'] = lst_raster
    my_inputs['lst_raster'] = lst_raster

    my_inputs['ndvi_raster'] = ndvi_raster
    my_inputs['ndwi_raster'] = ndwi_raster
    
    my_inputs['ndvi_qa'] = ndvi_qa
    my_inputs['ndwi_qa'] = ndwi_qa

    
    my_inputs['dt_raster'] = dt_raster
    my_inputs['tmax_raster'] = tmax_raster
    my_inputs['etr_raster'] = etr_raster

    my_inputs['global_water_raster'] = global_water_raster
    
    return(my_inputs)

    

    

In [5]:
my_inputs = my_favorite_rasters()

In [7]:
my_inputs

{'gold_standard': '/wsefs/pipeline/data_test_normToGrid/LS_eVSE_TEMP.2022.284-293.1KM.LST_TEMP.006.2022294042535.tif',
 'lst_raster': '/wsefs/pipeline/data_test_normToGrid/LS_eVSE_TEMP.2022.284-293.1KM.LST_TEMP.006.2022294042535.tif',
 'ndvi_raster': '/wsefs/pipeline/data_test_normToGrid/GN_eVSH_NDVI.2022.284-293.1KM.VI_NDVI.001.2022310010552.tif',
 'ndwi_raster': '/wsefs/pipeline/data_test_normToGrid/GN_eVSH_NDWI.2022.284-293.1KM.WI_NDWI.006.2022310173428.tif',
 'ndvi_qa': '/wsefs/pipeline/data_test_normToGrid/GN_eVSH_NDVI.2022.284-293.1KM.VI_QUAL.001.2022310010552.tif',
 'ndwi_qa': '/wsefs/pipeline/data_test_normToGrid/GN_eVSH_NDWI.2022.284-293.1KM.WI_QUAL.006.2022310173428.tif',
 'dt_raster': 's3://ws-in/ssebop/dT/dT_102.tif',
 'tmax_raster': 's3://ws-in/ssebop/air_temp/tmax_102.tif',
 'etr_raster': 's3://ws-in/ssebop/etr/Hobbins/etr_102.tif',
 'global_water_raster': 's3://ws-in/ssebop/masks/global_water_mask_inland.tif'}

In [8]:
s = rasterio.open('s3://ws-in/ssebop/masks/global_water_mask_inland.tif')

In [9]:
s.transform

Affine(0.0020810045, 0.0, -179.99999999990004,
       0.0, -0.0020810045, 83.62524685472488)

In [10]:
# dir(s)

In [11]:
s.dtypes

('uint16',)

In [12]:
s.units

(None,)

In [13]:
def get_src_dict(file_url):
    d={}
    src = rasterio.open(file_url)
    d['crs'] = src.crs.to_string()
    d['dtypes'] = src.dtypes
    d['height'] = src.height
    d['width'] = src.width
    d['bounds'] = src.bounds
    d['compression'] = src.compression
    d['transform'] = src.transform
    return(d)

In [14]:
def return_characteristics(the_inputs):
    #print(the_inputs.keys())
    d={}
    for ky in the_inputs.keys():
        d['name'] = ky
        d['file_url'] = the_inputs[ky]
        cool_dict = get_src_dict(d['file_url'])
        print(ky,cool_dict)

In [15]:
return_characteristics(my_inputs)

gold_standard {'crs': 'EPSG:4326', 'dtypes': ('uint16',), 'height': 14505, 'width': 37298, 'bounds': BoundingBox(left=-180.0, bottom=-60.002258848398924, right=180.00029303878546, top=80.0), 'compression': None, 'transform': Affine(0.009651999920606613, 0.0, -180.0,
       0.0, -0.009651999920606613, 80.0)}
lst_raster {'crs': 'EPSG:4326', 'dtypes': ('uint16',), 'height': 14505, 'width': 37298, 'bounds': BoundingBox(left=-180.0, bottom=-60.002258848398924, right=180.00029303878546, top=80.0), 'compression': None, 'transform': Affine(0.009651999920606613, 0.0, -180.0,
       0.0, -0.009651999920606613, 80.0)}
ndvi_raster {'crs': 'EPSG:4326', 'dtypes': ('int16',), 'height': 14505, 'width': 37298, 'bounds': BoundingBox(left=-180.0, bottom=-60.002258848398924, right=180.00029303878546, top=80.0), 'compression': None, 'transform': Affine(0.009651999920606613, 0.0, -180.0,
       0.0, -0.009651999920606613, 80.0)}
ndwi_raster {'crs': 'EPSG:4326', 'dtypes': ('int16',), 'height': 14505, 'width'

# Issues

basically we should create better reference iamges in s3.

- global_water_raster  - is huge and should be preprocessed at lower resolution and clipped to Kirk Evenson (Bounds)
    - global_water_raster - is not wide enuf -179.99 instead of 180.0002 - width will be 37297 insted of 37298
- ndwi_qa - likely needs to be in the list - just an omission - no biggie
- data types should standardize on int16 dtype
- compression and tiling would be nice to reduce disk usage

In [16]:
#! mkdir /wsefs/pipeline/data_temp

In [17]:
help(os.system)

Help on built-in function system in module posix:

system(command)
    Execute the command in a subshell.



In [18]:
# ! rio clip --help

In [19]:
my_inputs['gold_standard']

'/wsefs/pipeline/data_test_normToGrid/LS_eVSE_TEMP.2022.284-293.1KM.LST_TEMP.006.2022294042535.tif'

In [20]:
my_inputs

{'gold_standard': '/wsefs/pipeline/data_test_normToGrid/LS_eVSE_TEMP.2022.284-293.1KM.LST_TEMP.006.2022294042535.tif',
 'lst_raster': '/wsefs/pipeline/data_test_normToGrid/LS_eVSE_TEMP.2022.284-293.1KM.LST_TEMP.006.2022294042535.tif',
 'ndvi_raster': '/wsefs/pipeline/data_test_normToGrid/GN_eVSH_NDVI.2022.284-293.1KM.VI_NDVI.001.2022310010552.tif',
 'ndwi_raster': '/wsefs/pipeline/data_test_normToGrid/GN_eVSH_NDWI.2022.284-293.1KM.WI_NDWI.006.2022310173428.tif',
 'ndvi_qa': '/wsefs/pipeline/data_test_normToGrid/GN_eVSH_NDVI.2022.284-293.1KM.VI_QUAL.001.2022310010552.tif',
 'ndwi_qa': '/wsefs/pipeline/data_test_normToGrid/GN_eVSH_NDWI.2022.284-293.1KM.WI_QUAL.006.2022310173428.tif',
 'dt_raster': 's3://ws-in/ssebop/dT/dT_102.tif',
 'tmax_raster': 's3://ws-in/ssebop/air_temp/tmax_102.tif',
 'etr_raster': 's3://ws-in/ssebop/etr/Hobbins/etr_102.tif',
 'global_water_raster': 's3://ws-in/ssebop/masks/global_water_mask_inland.tif'}

In [21]:
# ! rio clip s3://ws-in/ssebop/masks/global_water_mask_inland.tif /wsefs/pipeline/data_temp/global_water_mask_inland.tif \
#     --like /wsefs/pipeline/data_test_normToGrid/LS_eVSE_TEMP.2022.284-293.1KM.LST_TEMP.006.2022294042535.tif \
#     --overwrite --co tiled=true --co blockxsize=256 --co blockysize=256 --co compress=LZW

In [22]:
! cat my_profile.json

defaults = {
        'driver': 'GTiff',
        'interleave': 'band',
        'tiled': True,
        'blockxsize': 256,
        'blockysize': 256,
        'compress': 'lzw',
        'nodata': 0,
        'dtype': uint16
    }

In [23]:
46553108272 / 1e9

46.553108272

In [24]:
# ! rio clip s3://ws-in/ssebop/masks/global_water_mask_inland.tif /wsefs/pipeline/data_temp/global_water_mask_inland.tif \
#     --bounds '[-180.0, -60.002258848398924, 180.00029303878546, 80.0]' --overwrite

In [25]:

'''
import rasterio
from rasterio.windows import from_bounds
from rasterio.enums import Resampling

(left, bottom, right, top) = [-180.0, -60.002258848398924, 180.00029303878546, 80.0]

filepath = my_inputs['global_water_raster']
with rasterio.open(filepath) as src:
    rst = src.read(1, window=from_bounds(left, bottom, right, top, src.transform))
'''

"\nimport rasterio\nfrom rasterio.windows import from_bounds\nfrom rasterio.enums import Resampling\n\n(left, bottom, right, top) = [-180.0, -60.002258848398924, 180.00029303878546, 80.0]\n\nfilepath = my_inputs['global_water_raster']\nwith rasterio.open(filepath) as src:\n    rst = src.read(1, window=from_bounds(left, bottom, right, top, src.transform))\n"

In [26]:
# dir(rst)

In [27]:
# rst.shape

In [28]:
# src.shape

In [29]:
sample_file = my_inputs['gold_standard']

In [30]:
with rasterio.open(sample_file) as src:

        out_meta = src.meta
        crs = out_meta['crs']
        transform = out_meta['transform']
        left = transform[2]
        top = transform[5]
        cols = out_meta['width']
        rows = out_meta['height']
        xres = transform[0]
        yres = transform[4]
        # return out_meta


In [31]:
cols

37298

In [32]:
cols= cols - 1 # ugly fudge for not big enuf data - see issues

In [33]:
cols

37297

In [34]:
from rasterio.enums import Resampling
from rasterio.vrt import WarpedVRT
from rasterio import shutil as rio_shutil

def warp_me(sample_file, in_file, out_file, out_datatype):
    
    with rasterio.open(sample_file) as src:

        out_meta = src.meta
        crs = out_meta['crs']
        transform = out_meta['transform']
        left = transform[2]
        top = transform[5]
        cols = out_meta['width']
        rows = out_meta['height']
        xres = transform[0]
        yres = transform[4]

    my_creation_options = {
        'compress':'LZW',
        'tiled':True, 
        'blockxsize':256, 
        'blockysize':256
    }

    rs = Resampling.nearest
    outdtype = out_datatype
    warpfile = in_file
    with rasterio.open(warpfile, 'r') as src:
        with WarpedVRT(src, resampling=rs,
                               crs=crs,
                               transform=transform,
                               height=rows,
                               width=cols,
                               dtype=outdtype) as vrt:
            outwarp=out_file
            rio_shutil.copy(vrt, outwarp, driver='GTiff', **my_creation_options)

In [35]:
my_inputs.keys()

dict_keys(['gold_standard', 'lst_raster', 'ndvi_raster', 'ndwi_raster', 'ndvi_qa', 'ndwi_qa', 'dt_raster', 'tmax_raster', 'etr_raster', 'global_water_raster'])

In [36]:
def get_dtype(file_name_key):
    int_list = ['ndvi_qa', 'ndwi_qa', 'global_water_raster']
    if any(file_name_key in s for s in int_list):
        out_data_type = 'uint16'
    else:
        out_data_type = 'float32'
    print(f'out_data_type:{out_data_type}')
    return out_data_type

In [37]:
ndata='/wsefs/pipeline/ndata'
sample_file = my_inputs['gold_standard']
for ky in my_inputs.keys():
    print(ky)
    input_file = my_inputs[ky]
    print(input_file)
    output_file = f'{ndata}/{ky}.tif'
    dtype = get_dtype(ky)
    warp_me(sample_file, input_file, output_file, dtype)

gold_standard
/wsefs/pipeline/data_test_normToGrid/LS_eVSE_TEMP.2022.284-293.1KM.LST_TEMP.006.2022294042535.tif
out_data_type:float32
lst_raster
/wsefs/pipeline/data_test_normToGrid/LS_eVSE_TEMP.2022.284-293.1KM.LST_TEMP.006.2022294042535.tif
out_data_type:float32
ndvi_raster
/wsefs/pipeline/data_test_normToGrid/GN_eVSH_NDVI.2022.284-293.1KM.VI_NDVI.001.2022310010552.tif
out_data_type:float32
ndwi_raster
/wsefs/pipeline/data_test_normToGrid/GN_eVSH_NDWI.2022.284-293.1KM.WI_NDWI.006.2022310173428.tif
out_data_type:float32
ndvi_qa
/wsefs/pipeline/data_test_normToGrid/GN_eVSH_NDVI.2022.284-293.1KM.VI_QUAL.001.2022310010552.tif
out_data_type:uint16
ndwi_qa
/wsefs/pipeline/data_test_normToGrid/GN_eVSH_NDWI.2022.284-293.1KM.WI_QUAL.006.2022310173428.tif
out_data_type:uint16
dt_raster
s3://ws-in/ssebop/dT/dT_102.tif
out_data_type:float32
tmax_raster
s3://ws-in/ssebop/air_temp/tmax_102.tif
out_data_type:float32
etr_raster
s3://ws-in/ssebop/etr/Hobbins/etr_102.tif
out_data_type:float32
global_w

In [38]:
# help(rio_shutil)

In [39]:
import glob

In [40]:
ndata = '/wsefs/pipeline/ndata'

my_outs = glob.glob(f'{ndata}/*.tif')

In [41]:
my_outs

['/wsefs/pipeline/ndata/lst_raster.tif',
 '/wsefs/pipeline/ndata/ndwi_qa.tif',
 '/wsefs/pipeline/ndata/global_water_raster.tif',
 '/wsefs/pipeline/ndata/etr_raster.tif',
 '/wsefs/pipeline/ndata/ndvi_raster.tif',
 '/wsefs/pipeline/ndata/dt_raster.tif',
 '/wsefs/pipeline/ndata/ndvi_qa.tif',
 '/wsefs/pipeline/ndata/ndwi_raster.tif',
 '/wsefs/pipeline/ndata/tmax_raster.tif',
 '/wsefs/pipeline/ndata/gold_standard.tif',
 '/wsefs/pipeline/ndata/etf_test.tif']

In [42]:
for i in my_outs:
    print(i)
    d = get_src_dict(i)
    print(d)
 

/wsefs/pipeline/ndata/lst_raster.tif
{'crs': 'EPSG:4326', 'dtypes': ('float32',), 'height': 14505, 'width': 37298, 'bounds': BoundingBox(left=-180.0, bottom=-60.002258848398924, right=180.00029303878546, top=80.0), 'compression': <Compression.lzw: 'LZW'>, 'transform': Affine(0.009651999920606613, 0.0, -180.0,
       0.0, -0.009651999920606613, 80.0)}
/wsefs/pipeline/ndata/ndwi_qa.tif
{'crs': 'EPSG:4326', 'dtypes': ('uint16',), 'height': 14505, 'width': 37298, 'bounds': BoundingBox(left=-180.0, bottom=-60.002258848398924, right=180.00029303878546, top=80.0), 'compression': <Compression.lzw: 'LZW'>, 'transform': Affine(0.009651999920606613, 0.0, -180.0,
       0.0, -0.009651999920606613, 80.0)}
/wsefs/pipeline/ndata/global_water_raster.tif
{'crs': 'EPSG:4326', 'dtypes': ('uint16',), 'height': 14505, 'width': 37298, 'bounds': BoundingBox(left=-180.0, bottom=-60.002258848398924, right=180.00029303878546, top=80.0), 'compression': <Compression.lzw: 'LZW'>, 'transform': Affine(0.009651999920

AttributeError: 'NoneType' object has no attribute 'to_string'